enhancements/console/customize-perspectives.md

```diff
@@ -0,0 +1,278 @@
+
+Additional perspectives can be dynamically added with the 
+[Perspective extension](https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/extensions/perspectives.ts).
+
+### Goals
+
+- Allow cluster admins to hide the user perspectives.
+- Provide a way that cluster admins can do this based on resource permissions (RBAC).
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+We can extend the existing `operator.openshift.io/v1` / `Console` CRD. 
+We can add 
+`spec.customization.perspectives` which can have list of customization details.
+
+### Workflow Description
+
+To disable perspective for users with limited permissions we need a list of options with a perspective id and the required permissions.
+
+```text
+├── Spec
+│   ├── customization
+|       ├── perspectives 
+|           ├── id
+|           ├── requiresAccessReview
+|               ├── kind
+|               ├── verb
+└── ...
+```
+
+This might be extendable in the future if we want replace (hide one, show another) a perspective based on a resource permissions:
+
+```text
+├── Spec
+│   ├── customization
+|       ├── perspectives 
+|           ├── id
+|           ├── requiresAccessReview
+|           |   ├── kind
+|           |   ├── verb
+|           ├── missingAccessReview
+|               ├── kind
+|               ├── verb
+└── ...
+```
+
+Each perspective conforms to the following schema:
+
+- `id: string` , defines the id of perspective to disable. Incorrect or unknown ids will be ignored.
+
+// +optional
+- `requiresAccessReview: authV1.ResourceAttributes[]` , defines a list of permission checks. The perspective will only be shown when all checks are successful.
+
+// +optional
+- `missingAccessReview: authV1.ResourceAttributes[]` , defines a list of permission checks. The perspective will only be shown when at least one check fails.
+
+### API Extensions
+
+Extend the existing `operator.openshift.io/v1` / `Console` CRD to add `spec.customization.perspectives` which can have list of customization details.
+
+So the final yaml should look like this one:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata: 
+  name: cluster
+  ...
+spec:
+  customization:
+    perspectives:
+      - id: admin
+        requiresAccessReview:
+          - kind: namespaces
+            verb: list
+      - id: dev
+        state: Enabled/Disabled
+```
+
+Example yaml based on resource permission:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata: 
+  name: cluster
+  ...
+spec:
+  customization:
+    perspectives:
+      - id: admin
+        requiresAccessReview:
+          - kind: namespaces
+            verb: list
+      - id: minimal-custom-admin
+        missingAccessReview:
+          - kind: namespaces
+            verb: list
+```
+### Risks and Mitigations
+
+**Default perspective**: When all the perspectives are hidden from a user or for all users, we always show Admin perspective by default and if admin has to 
```

Comment:
I'm guessing that the motivation behind this is a security one? This seems like a security flaw, how will you mitigate it?

Reply:
It's not a security issue. It's more about what is the behavior if an admin tries to disable all perspectives? You have to have at least one perspective to see anything in the console left nav.

The perspectives themselves hide/show elements based on the user's permissions (and of course the API server enforces RBAC as well).

Reply:
Ack ok, so the motivation here is that, if an admin misconfigures the perspectives, they need access to the admin perspective to rectify that issue, so we show the admin perspective by default? 

Perhaps the console could show a warning when that happens?

Reply:
This comment is still outstanding